### PR TITLE
Llama2 TPU optimizations

### PR DESCRIPTION
**Pull request by miladm**
This PR migrates LLaMA optimizations to LLaMA2 codebase and performs additional optimizations to align the the two versions of the code.

### CODE REVIEW

File: llama/generation.py [hunk 1]
```diff
@@ -102,68 +104,124 @@ def build(
...
+
+        # TODO: check if we need to optimize this block - llama1 has a simpler logic
         min_prompt_len = min(len(t) for t in prompt_tokens)
         max_prompt_len = max(len(t) for t in prompt_tokens)
         assert max_prompt_len <= params.max_seq_len
         total_len = min(params.max_seq_len, max_gen_len + max_prompt_len)
 
         pad_id = self.tokenizer.pad_id
-        tokens = torch.full((bsz, total_len), pad_id, dtype=torch.long, device="cuda")
+        tokens = torch.full((bsz, total_len), pad_id, dtype=torch.long)
```
Comment (Liyang90) [inline llama/generation.py]: This was modified in LLaMA 1 to avoid dynamic shape from different inputs.

File: llama/generation.py [hunk 2]
```diff
@@ -102,68 +104,124 @@ def build(
...
-            if logprobs:
-                token_logprobs[:, prev_pos + 1 : cur_pos + 1] = -F.cross_entropy(
-                    input=logits.transpose(1, 2),
-                    target=tokens[:, prev_pos + 1 : cur_pos + 1],
-                    reduction="none",
-                    ignore_index=pad_id,
+        xm.mark_step(wait=True)
+
+        decoding_start_time = time.time()
+        for cur_pos in range(min_prompt_len, total_len): #TODO: drop dependency on cur_pos. min_prompt_len used to be start_pos
```
Comment (Liyang90) [inline llama/generation.py]: `cur_pos` should match `cur_pos_tensor`.

File: llama/model.py [hunk 3]
```diff
@@ -87,100 +95,114 @@ def repeat_kv(x: torch.Tensor, n_rep: int) -> torch.Tensor:
...
-        self.n_kv_heads = args.n_heads if args.n_kv_heads is None else args.n_kv_heads
-        model_parallel_size = fs_init.get_model_parallel_world_size()
-        self.n_local_heads = args.n_heads // model_parallel_size
-        self.n_local_kv_heads = self.n_kv_heads // model_parallel_size
-        self.n_rep = self.n_local_heads // self.n_local_kv_heads
+        #NEW ADDITION
+        #self.n_kv_heads = args.n_heads if args.n_kv_heads is None else args.n_kv_heads
+        #model_parallel_size = fs_init.get_model_parallel_world_size()
+        #self.n_local_heads = args.n_heads // model_parallel_size
+        #self.n_local_kv_heads = self.n_kv_heads // model_parallel_size
```
Comment (Liyang90) [inline llama/model.py:105]: "grouped-query attention" is the major LLaMA 2 difference from LLaMA 1. Without it, there is no need to adapt optimization to LLaMA 2, because the code would be same.

File: llama/model.py [hunk 4]
```diff
@@ -87,100 +95,114 @@ def repeat_kv(x: torch.Tensor, n_rep: int) -> torch.Tensor:
...
 
+        init_method = lambda x: x
+
         self.wq = ColumnParallelLinear(
             args.dim,
             args.n_heads * self.head_dim,
             bias=False,
             gather_output=False,
-            init_method=lambda x: x,
+            init_method=lambda x: x, #THIS CALL IS MADE SIMPLER: WORLD_SIZE, RANK, GROUP
```
Comment (Liyang90) [inline llama/model.py]: The `world_size`, `rank`, `group` needs to be passed in from here to allow us arbitrarily instantiate a shard of the model to do resharding. It can not be simplified.

File: llama/model.py [hunk 5]
```diff
@@ -190,99 +212,167 @@ def __init__(
...
-            ffn_dim_multiplier=args.ffn_dim_multiplier,
+            world_size=world_size, rank=rank, groups=groups,
+            ffn_dim_multiplier=args.ffn_dim_multiplier
         )
-        self.layer_id = layer_id
-        self.attention_norm = RMSNorm(args.dim, eps=args.norm_eps)
-        self.ffn_norm = RMSNorm(args.dim, eps=args.norm_eps)
+        self.layer_id = layer_id #NEW LINE
+        self.attention_norm = RMSNorm(args.dim, eps=args.norm_eps) #NEW LINE
+        self.ffn_norm = RMSNorm(args.dim, eps=args.norm_eps) #NEW LINE
```
Comment (Liyang90) [inline llama/model.py]: What does `#NEW LINE` mean? They seem to be same in LLaMA 1.

File: llama/generation.py [hunk 6]
```diff
@@ -102,68 +104,124 @@ def build(
...
+            token_logprobs[:, prev_pos + 1 : cur_pos + 1] = -F.cross_entropy(
+                input=logits.transpose(1, 2),
+                target=tokens[:, prev_pos + 1 : cur_pos + 1],
+                reduction="none",
+                ignore_index=pad_id,
+            )
+        #####
+
+        if temperature > 0:
+            probs = torch.softmax(logits[:, -1] / temperature, dim=-1) #TODO: eval the perf impact of logits vs. logits[:,-1]
```
Comment (Liyang90) [inline llama/generation.py]: Cannot simply take the last position of logits. Need to use the `output_pos_tensor`

File: llama/model.py [hunk 7]
```diff
@@ -190,99 +212,167 @@ def __init__(
...
-        for layer in self.layers:
-            h = layer(h, start_pos, freqs_cis, mask)
+        new_cache_kvs = []
+        for layer, cache_kv in zip(self.layers, cache_kvs):
+            h, new_cache_kv = layer(h, freqs_cis, mask, input_idexes, cache_kv)
+            new_cache_kvs.append(new_cache_kv)
         h = self.norm(h)
-        output = self.output(h).float()
-        return output
+        h = h.index_select(1, output_idex - input_idexes[0]).squeeze(dim=1)
```
Comment (Liyang90) [inline llama/model.py:371]: LLaMA 2 changed to output logits at all positions, so this needs to be changed accordingly.

**Comment by Liyang90** [comment]
[TODO] The optimization in LLaMA 1 for long prompt and dynamic temperature and top p values are not included.

File: llama/generation.py [hunk 8]
```diff
@@ -292,12 +343,34 @@ def chat_completion(
             for t in generation_tokens
         ]
 
+    def prompt_completion(
```
Comment (Liyang90) [inline llama/generation.py]: Can we just use `text_completion`? This seems to do the same thing as `text_completion`.

Comment (miladm) [reply llama/generation.py]: correct - it's pretty much the `decode` call that's additional which is good to incorporate given it is no longer in the `generate()` method - I will drop this and repurpose calls to use `text_completion`

File: llama/generation.py (hunk 2 above)
Comment (miladm) [reply llama/generation.py]: replaced this loop with what we had in llama1

File: llama/generation.py (hunk 1 above)
Comment (miladm) [reply llama/generation.py]: optimized now.

File: llama/model.py (hunk 3 above)
Comment (miladm) [reply llama/model.py:105]: the grouped-query attention is purposely commented - as discussed let's address that in the follow up PR.

Comment (miladm) [reply llama/model.py:105]: in that PR, we bring our llama1 optimized kv-cache data structure flow, but use grouped-query attention. works?

File: llama/model.py (hunk 4 above)
Comment (miladm) [reply llama/model.py]: yes - all needed parameters were added while cleaning the `model.py` comments.

File: llama/model.py (hunk 5 above)
Comment (miladm) [reply llama/model.py]: lol - don't recall - they indeed exist in the previous generation

**Comment by miladm** [comment]
@Liyang90 will submit a PR to cover the kv-cache optimization. hope this helps unblock.

File: llama/generation.py [hunk 9]
```diff
@@ -102,17 +104,57 @@ def build(
     def __init__(self, model: Transformer, tokenizer: Tokenizer):
         self.model = model
         self.tokenizer = tokenizer
+        self._generate_one_token_fn = self._generate_one_token
+        self._generate_one_token_fn = torch.compile(self._generate_one_token_fn,
+                                                    backend="torchxla_trace_once", fullgraph=True)
+
+
+    def _generate_one_token(self, tokens, input_tokens, input_text_mask, cur_pos_tensor, 
+                            input_pos_tensor, output_pos_tensor, cache_kvs, temperature, top_p, logprobs, prev_pos):
```
Comment (Liyang90) [inline llama/generation.py:113]: `prev_pos` this argument here would make dynamo retrace and recompile for every token generation.

Comment (miladm) [reply llama/generation.py:113]: we should get rid of this variable - it's not really used if you look at the function details.

Comment (Liyang90) [reply llama/generation.py:113]: I fixed this in PR:https://github.com/pytorch-tpu/llama/pull/30
